dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionary.java

```diff
@@ -17,156 +17,342 @@
-   * @see
-   * org.icgc.dcc.submission.validation.key.core.KVDictionary#getErrorFieldNames(org.icgc.dcc.submission.validation.
-   * key.core.KVFileType, org.icgc.dcc.submission.validation.key.core.KVErrorType)
-   */
+  public List<String> getPrimaryKeyNames(@NonNull KVFileType fileType) {
+    val fileSchema = getFileSchema(fileType);
+
+    return fileSchema.getUniqueFields().stream()
+        .sorted()
+        .collect(toImmutableList());
+  }
+
   @Override
-  public List<String> getErrorFieldNames(KVFileType fileType, KVErrorType errorType) {
-    // TODO Auto-generated method stub
-    return null;
-  }
-
-  /*
-   * (non-Javadoc)
-   * 
-   * @see
-   * org.icgc.dcc.submission.validation.key.core.KVDictionary#getPrimaryKeyNames(org.icgc.dcc.submission.validation.
-   * key.core.KVFileType)
-   */
+  public Iterable<KVFileType> getTopologicallyOrderedFileTypes() {
+    return KVDynamicDictionaryHelper.getTopologicallyOrderedFileTypes(dictionary);
+  }
+
   @Override
-  public List<String> getPrimaryKeyNames(KVFileType fileType) {
-    // TODO Auto-generated method stub
-    return null;
+  public Collection<KVFileType> getParents(@NonNull KVFileType fileType) {
+    return dictionary.getParents(fileType.getFileType()).stream()
+        .map(KVFileType::from)
+        .collect(toImmutableList());
   }
 
-  /*
-   * (non-Javadoc)
-   * 
-   * @see org.icgc.dcc.submission.validation.key.core.KVDictionary#getSurjectionForeignKeyNames(org.icgc.dcc.submission.
-   * validation.key.core.KVFileType)
-   */
   @Override
-  public List<String> getSurjectionForeignKeyNames(KVFileType fileType) {
-    // TODO Auto-generated method stub
-    return null;
-  }
-
-  /*
-   * (non-Javadoc)
-   * 
-   * @see
-   * org.icgc.dcc.submission.validation.key.core.KVDictionary#getReferencingFileType(org.icgc.dcc.submission.validation
-   * .key.core.KVFileType)
-   */
+  public boolean hasChildren(@NonNull KVFileType fileType) {
+    return dictionary.getFiles().stream()
+        .flatMap(file -> file.getRelations().stream())
+        .map(relation -> relation.getOtherFileType())
+        .map(KVFileType::from)
+        .anyMatch(ft -> ft == fileType);
+  }
+
   @Override
-  public KVFileType getReferencingFileType(KVFileType fileType) {
-    // TODO Auto-generated method stub
-    return null;
+  public Collection<KVFileType> getSurjectiveReferencedTypes(@NonNull KVFileType fileType) {
+    val fileSchema = getFileSchema(fileType);
+
+    return fileSchema.getRelations().stream()
+        .filter(Relation::isBidirectional)
+        .map(Relation::getOtherFileType)
+        .map(KVFileType::from)
+        .collect(toImmutableList());
+  }
+
+  private Map<KVFileType, RowConditionEvaluator> createConditionEvaluators(@NonNull KVFileType fileType) {
+    val conditions = resolveConditions(fileType);
+    val conditionFields = getTypeFields(fileType, dictionary);
+
+    return conditions.entrySet().stream()
+        .collect(toImmutableMap(entry -> entry.getKey(),
```

Comment:
There is another version of `toImmutableMap` that doesn't require this first argument as it is almost always this pattern.